### PR TITLE
NIFI-10532 ensuring client gets reset if any of the key values host/p…

### PR DESCRIPTION
**Pull request by joewitt**
…ort/user/pw change on a per ff basis

# Summary

[NIFI-00000](https://issues.apache.org/jira/browse/NIFI-00000)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by joewitt** [comment]
need to do some manual testing now on before/after.  Have only walked through the code and made suggeste change.  Will share details soon

**Comment by joewitt** [comment]
Verified behavior before was not correct.  Verified behavior after is working as expected.  Verified logic in provided example from NII-10532 before/after specifically for FTP processors.  Tested behavior of PutFTP, GetFTP, ListFTP, FetchFTP. 

Need to check SFTP now as well but the exact same change was made.

**Comment by joewitt** [comment]
The issue doesnt appear to occur on the previous version though the change should close the risk and also tests to work well.

Testing did reveal that file listing logic doesn't work in the case parameters come from flow files as is needed with teh PutSFTP case even to check for missing directories.  That can be addressed in a later issue as it isn't a risk of sending data to the wrong destination.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/SFTPTransfer.java [hunk 1]
```diff
@@ -580,12 +587,23 @@ private String getMessage(final SFTPException e) {
     }
 
     protected SFTPClient getSFTPClient(final FlowFile flowFile) throws IOException {
+        final String evaledHostname = ctx.getProperty(HOSTNAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String evaledPort = ctx.getProperty(PORT).evaluateAttributeExpressions(flowFile).getValue();
+        final String evaledUsername = ctx.getProperty(PORT).evaluateAttributeExpressions(flowFile).getValue();
+        final String evaledPassword = ctx.getProperty(PORT).evaluateAttributeExpressions(flowFile).getValue();
+        final String evaledPrivateKeyPath = ctx.getProperty(PORT).evaluateAttributeExpressions(flowFile).getValue();
+        final String evaledPrivateKeyPassphrase = ctx.getProperty(PORT).evaluateAttributeExpressions(flowFile).getValue();
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/SFTPTransfer.java:595]: It looks like the property descriptor `PORT` reference needs to be replaced with the appropriate property descriptor for the username, password, and private key values.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for addressing this problem @joewitt! I noted one issue with `SFTPTransfer` property descriptor references, but the approach look straightforward.

Although it would add some complexity, it might be worth considering creating a hash of all relevant properties and use that for comparison, as opposed to keeping another direct copy of the credentials in memory.

**Comment by joewitt** [comment]
haha - yeah already pushed (force) the PORT PORT PORT thing.

I also thought about the hashing which would be a fine approach.  But I'd do that as part of what appears to be a lot of surgery needed in this.  It has grown in purposes/reuse for other patterns and frankly could use a substantial house cleaning.

**Comment by exceptionfactory** [comment]
Thanks @joewitt, agreed, re-evaluating the approach at a higher level sounds like a good topic for a separate issue.

**Review by arpadboda — APPROVED** [review]

**Comment by markap14** [comment]
Thanks @joewitt for the fix. And to @exceptionfactory and @arpadboda for reviewing as well. Code changes make sense (though I agree it would make sense to cleanup in the future). Was able to do some testing. All looks good to me as well. Will merge to main.

**Pull request merged**